### ISSUE TITLE
Outgoing HTTP request never completes

### ISSUE THREAD

**Issue by Ne4to** [dotnet/runtime#50169]
It is continue of #40541 (I was not able to reopen it or add a comment, feel free to merge two issues). Now it is reproduced with the latest version of .NET 5.0

### Description
`dotnet-dump analyze` for two memory dumps shows the same _async stack_. These dumps were made with 35 minutes interval. Using following sequence of command I found request's Uri. It contains an unique Id. Logs from other party for this Id show successfull response with 200 status code executed in less than a second.

`AsyncStateMachineBox` -> `StateMachine` -> `<>4__this (HttpConnection)` -> `_currentRequest (HttpRequestMessage)` -> `_requestUri`
```
> dumpasync -stacks
         Address               MT     Size      State Description
00007f1190624d18 00007f12ea20e038      128          0 System.Net.Http.HttpConnection+<FillAsync>d__94
Async "stack":
.00007f1190624d98 (1) System.Net.Http.HttpConnection+ChunkedEncodingReadStream+<CopyToAsyncCore>d__10
..00007f1190624e38 (0) System.Net.Http.HttpConnectionResponseContent+<SerializeToStreamAsync>d__6
...00007f1190624ed0 (0) System.Net.Http.HttpContent+<LoadIntoBufferAsyncCore>d__63
....00007f1190623ee0 (0) <error>
.....00007f1190623fa8 (0) <error>
......00007f1190624050 (0) <error>
.......00007f11906241a0 (0) <error>
........00007f1190624230 System.Threading.Tasks.Task+SetOnInvokeMres

> do 00007f1190624d18
Name:        System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib],[System.Net.Http.HttpConnection+<FillAsync>d__94, System.Net.Http]]
MethodTable: 00007f12ea20e038
EEClass:     00007f12ea21a7f0
Size:        128(0x80) bytes
File:        /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Private.CoreLib.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007f12e65bb158  4000b80       30         System.Int32  1 instance                0 m_taskId
00007f12e6671a30  4000b81        8      System.Delegate  0 instance 0000000000000000 m_action
00007f12e65b0c00  4000b82       10        System.Object  0 instance 0000000000000000 m_stateObject
00007f12e75f60f0  4000b83       18 ...sks.TaskScheduler  0 instance 0000000000000000 m_taskScheduler
00007f12e65bb158  4000b84       34         System.Int32  1 instance         33555456 m_stateFlags
00007f12e65b0c00  4000b85       20        System.Object  0 instance 00007f1190624d98 m_continuationObject
00007f12e9968600  4000b89       28 ...tingentProperties  0 instance 0000000000000000 m_contingentProperties
00007f12e65bb158  4000b7f      8fc         System.Int32  1   static                0 s_taskIdCounter
00007f12e65b0c00  4000b86      8d8        System.Object  0   static 00007f1190030c08 s_taskCompletionSentinel
00007f12e65b7138  4000b87      900       System.Boolean  1   static                0 s_asyncDebuggingEnabled
0000000000000000  4000b88      8e0                       0   static 0000000000000000 s_currentActiveTasks
00007f12e75f4ac8  4000b8a      8e8 ...Tasks.TaskFactory  0   static 00007f1190030c20 <Factory>k__BackingField
00007f12e75f5908  4000b8b      8f0 ...Private.CoreLib]]  0   static 00007f1190030c48 s_cachedCompleted
00007f12e75f3d40  4000b8c      8f8 ...g.ContextCallback  0   static 00007f1190030c90 s_ecCallback
00007f12e67088c0  4000b7e       40 ...eading.Tasks.Task  0 TLstatic  t_currentTask
    >> Thread:Value 1:0000000000000000 d:00007f119024dd08 f:00007f1190030d58 10:00007f119031ba20 11:00007f11900e4e30 21e:0000000000000000 229:0000000000000000 22a:0000000000000000 22b:0000000000000000 <<
00007f12e75f53f8  4000b3b       38 ...ks.VoidTaskResult  1 instance 00007f1190624d50 m_result
0000000000000000  4000b3c        8                       0   static dynamic statics NYI                 s_Factory
00007f12e6c73710  4001069       40        System.Action  0 instance 0000000000000000 _moveNextAction
00007f12ea20d820  400106a       50 ...+<FillAsync>d__94  1 instance 00007f1190624d68 StateMachine
00007f12e6709448  400106b       48 ....ExecutionContext  0 instance 00007f11903de2a8 Context
00007f12e75f3d40  4001068        8 ...g.ContextCallback  0   static dynamic statics NYI                 s_callback

> dumpvc 00007f12ea20d820 00007f1190624d68
Name:        System.Net.Http.HttpConnection+<FillAsync>d__94
MethodTable: 00007f12ea20d820
EEClass:     00007f12ea21a330
Size:        56(0x38) bytes
File:        /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Net.Http.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007f12e65bb158  4000485        8         System.Int32  1 instance                0 <>1__state
00007f12ea184980  4000486       10 ...TaskMethodBuilder  1 instance 00007f1190624d78 <>t__builder
00007f12ea11d348  4000487        0 ...tp.HttpConnection  0 instance 00007f11904f1c78 <>4__this
00007f12e65b7138  4000488        c       System.Boolean  1 instance                1 async
00007f12ea1f5c38  4000489       18 ...Private.CoreLib]]  1 instance 00007f1190624d80 <>u__1

> do 00007f11904f1c78
Name:        System.Net.Http.HttpConnection
MethodTable: 00007f12ea11d348
EEClass:     00007f12ea10aa90
Size:        168(0xa8) bytes
File:        /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Net.Http.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007f12e6677a60  40006cb        8        System.String  0 instance 00007f1190624370 _lastDateHeaderValue
00007f12e6677a60  40006cc       10        System.String  0 instance 00007f11904f4420 _lastServerHeaderValue
00007f12e65bc4d8  40006cd       18         System.Int64  1 instance       1661136775 <CreationTickCount>k__BackingField
00007f12ea1122d8  400038b       20 ...ttpConnectionPool  0 instance 00007f119040c9b0 _pool
00007f12ea128290  400038c       28 ...et.Sockets.Socket  0 instance 00007f11904f2ed0 _socket
00007f12e758da38  400038d       30     System.IO.Stream  0 instance 00007f11904f1c48 _stream
00007f12ea120e30  400038e       38 ....TransportContext  0 instance 0000000000000000 _transportContext
00007f12ea1f3e38  400038f       40 ...System.Net.Http]]  0 instance 00007f11904f41d0 _weakThisRef
00007f12ea012260  4000390       48 ...ttpRequestMessage  0 instance 00007f1190623938 _currentRequest
00007f12e67007b0  4000391       50        System.Byte[]  0 instance 00007f11904f1d20 _writeBuffer
00007f12e65bb158  4000392       68         System.Int32  1 instance                0 _writeOffset
00007f12e65bb158  4000393       6c         System.Int32  1 instance            16379 _allowedReadLineBytes
00007f12e6693bb8  4000394       58      System.String[]  0 instance 00007f11904f41e8 _headerValues
00007f12ea11d230  4000395       88 ...Private.CoreLib]]  1 instance 00007f11904f1d00 _readAheadTask
00007f12e65bb158  4000396       70         System.Int32  1 instance                0 _readAheadTaskLock
00007f12e67007b0  4000397       60        System.Byte[]  0 instance 00007f11904f31b8 _readBuffer
00007f12e65bb158  4000398       74         System.Int32  1 instance                0 _readOffset
00007f12e65bb158  4000399       78         System.Int32  1 instance                0 _readLength
00007f12e65b7138  400039a       80       System.Boolean  1 instance                0 _inUse
00007f12e65b7138  400039b       81       System.Boolean  1 instance                0 _canRetry
00007f12e65b7138  400039c       82       System.Boolean  1 instance                0 _connectionClose
00007f12e65bb158  400039d       7c         System.Int32  1 instance                0 _disposed
00007f12e67007b0  4000384      1f0        System.Byte[]  0   static 00007f11901112c8 s_contentLength0NewlineAsciiBytes
00007f12e67007b0  4000385      1f8        System.Byte[]  0   static 00007f11901112f8 s_spaceHttp10NewlineAsciiBytes
00007f12e67007b0  4000386      200        System.Byte[]  0   static 00007f1190111320 s_spaceHttp11NewlineAsciiBytes
00007f12e67007b0  4000387      208        System.Byte[]  0   static 00007f1190111348 s_httpSchemeAndDelimiter
00007f12e67007b0  4000388      210        System.Byte[]  0   static 00007f1190111368 s_http1DotBytes
00007f12e65bce98  4000389      2b0        System.UInt64  1   static 3471766442030158920 s_http10Bytes
00007f12e65bce98  400038a      2b8        System.UInt64  1   static 3543824036068086856 s_http11Bytes

> do 00007f1190623938
Name:        System.Net.Http.HttpRequestMessage
MethodTable: 00007f12ea012260
EEClass:     00007f12ea020398
Size:        80(0x50) bytes
File:        /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Net.Http.dll
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007f12e65bb158  40002d2       38         System.Int32  1 instance                0 _sendStatus
00007f12e9f727b0  40002d3        8 ...t.Http.HttpMethod  0 instance 00007f11906236b8 _method
00007f12e7fe3520  40002d4       10           System.Uri  0 instance 00007f11906236e0 _requestUri
00007f12ea013758  40002d5       18 ...ttpRequestHeaders  0 instance 00007f1190623988 _headers
00007f12e6ded440  40002d6       20       System.Version  0 instance 00007f11900eaf70 _version
00007f12e78ee2c8  40002d7       3c         System.Int32  1 instance                0 _versionPolicy
00007f12e9f74728  40002d8       28 ....Http.HttpContent  0 instance 0000000000000000 _content
00007f12e65b7138  40002d9       40       System.Boolean  1 instance                0 _disposed
0000000000000000  40002da       30 ...ttpRequestOptions  0 instance 0000000000000000 _options

> do 00007f11906236e0
Name:        System.Uri
MethodTable: 00007f12e7fe3520
EEClass:     00007f12e7f89f10
Size:        56(0x38) bytes
File:        /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Private.Uri.dll
```

### Configuration
Docker image based on mcr.microsoft.com/dotnet/aspnet:5.0 running in a Kubernetes cluster
```
> kubectl version

Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"1dd5338295409edcfff11505e7bb246f0d325d15", GitTreeState:"clean", BuildDate:"2021-01-13T13:23:52Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.12-gke.1210", GitCommit:"199a41188dc0ca5d6d95b1cc7e8ba96e05f9dd0a", GitTreeState:"clean", BuildDate:"2021-02-05T18:03:16Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
```
From a container:
```
> dotnet --info

Host (useful for support):
  Version: 5.0.4
  Commit:  f27d337295

.NET SDKs installed:
  No SDKs were found.

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.4 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.4 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

> cat /etc/os-release

PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

### Regression?

Originally it was found in .Net Core 2.2 (see #40541)

**Comment by stephentoub** [comment]
From @geoffkizer in https://github.com/dotnet/runtime/issues/40541#issuecomment-670636352:

> The HttpClient code is expecting more data on the chunked-encoded response stream. It could be a bug, but it seems more likely that the server is sending an invalid response, e.g. not properly terminating the chunked encoding.

> I'd suggest looking at what your server is actually sending and see if it looks correct or not.

**Comment by Ne4to** [comment]
I will try to run `tcpdump` in a container. 
But in case of network issues, shouldn't the operation fails with some timeout exception?

**Comment by stephentoub** [comment]
> But in case of network issues, shouldn't the operation fails with some timeout exception?

It depends on the code you're using.  If you're using HttpClient without having changed the Timeout property, and you're using GetStringAsync/GetByteArrayAsync, then it should timeout after ~100 seconds.

**Comment by davidfowl** [comment]
> SetOnInvokeMres 

Something is doing a .Result or a .Wait. Can you share the callstacks?

**Comment by Ne4to** [comment]
> It depends on the code you're using.

It's complicated 😢. It is generated code based on an API contract and it is not using HttpClient, it uses HttpClientHandler directly (don't ask me why 😹). Internally main part looks like this:
```
HttpResponseMessage response = await _httpClientHandler.SendAsync(request, ct).ConfigureAwait(false);
if (response.Content != null)
  await response.Content.LoadIntoBufferAsync().ConfigureAwait(false);
return response;
```

> Something is doing a .Result or a .Wait. Can you share the callstacks?

It actually calls `.GetAwaiter().GetResult()` on the top level. I understand that we shound not use sync over async, it is some legacy code, I can try to update it to `async-await` if it is an issue here (even in .NET 5).

```
> clrstack -all
OS Thread Id: 0x1
        Child SP               IP Call Site
00007FFD5E4F77A0 00007f136080400c [HelperMethodFrame_1OBJ: 00007ffd5e4f77a0] System.Threading.Monitor.ObjWait(Boolean, Int32, System.Object)
00007FFD5E4F78D0 00007F12E616B27B System.Threading.ManualResetEventSlim.Wait(Int32, System.Threading.CancellationToken)
00007FFD5E4F7950 00007F12E617F260 System.Threading.Tasks.Task.SpinThenBlockingWait(Int32, System.Threading.CancellationToken)
00007FFD5E4F79B0 00007F12E617F0FB System.Threading.Tasks.Task.InternalWaitCore(Int32, System.Threading.CancellationToken)
00007FFD5E4F7A00 00007F12E61E8D96 System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
00007FFD5E4F7A20 00007F12E61E8D49 System.Runtime.CompilerServices.TaskAwaiter.GetResult()
00007FFD5E4F7A30 00007F12E65F63E4 APPNAME.Program.<Main>(System.String[])
OS Thread Id: 0xb
        Child SP               IP Call Site
00007F135A384D30 00007f13608043f9 [DebuggerU2MCatchHandlerFrame: 00007f135a384d30]
OS Thread Id: 0xc
        Child SP               IP Call Site
OS Thread Id: 0xd
        Child SP               IP Call Site
00007F12E5882D10 00007f13608043f9 [HelperMethodFrame_1OBJ: 00007f12e5882d10] System.Threading.Monitor.ObjWait(Boolean, Int32, System.Object)
00007F12E5882E40 00007F12EB0E3FCF /app/Quartz.dll!Unknown
00007F12E58835F0 00007F12EB0DD5E4 /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Private.CoreLib.dll!Unknown
00007F12E5883660 00007F12EB0DD4FB /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Private.CoreLib.dll!Unknown
00007F12E5883680 00007F12EB0DD4B1 /app/Quartz.dll!Unknown
00007F12E5883760 00007F12E61773E1 System.Threading.Tasks.Task`1[[System.__Canon, System.Private.CoreLib]].InnerInvoke()
00007F12E5883780 00007F12E6184891 System.Threading.Tasks.Task+<>c.<.cctor>b__277_0(System.Object)
00007F12E5883790 00007F12E6169992 System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(System.Threading.Thread, System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
00007F12E58837D0 00007F12E617E7B9 System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
00007F12E5883850 00007F12E617E598 System.Threading.Tasks.Task.ExecuteEntryUnsafe(System.Threading.Thread)
00007F12E5883870 00007F12E617E537 System.Threading.Tasks.Task.ExecuteFromThreadPool(System.Threading.Thread)
00007F12E5883880 00007F12E6172B73 System.Threading.ThreadPoolWorkQueue.Dispatch()
00007F12E5883900 00007F12E6161E67 System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
00007F12E5883C80 00007f135fab9b47 [DebuggerU2MCatchHandlerFrame: 00007f12e5883c80]
OS Thread Id: 0xf
        Child SP               IP Call Site
00007F12E5082570 00007f1360807544 [InlinedCallFrame: 00007f12e5082570] Interop+Sys.Read(System.Runtime.InteropServices.SafeHandle, Byte*, Int32)
00007F12E5082570 00007f12e660fadb [InlinedCallFrame: 00007f12e5082570] Interop+Sys.Read(System.Runtime.InteropServices.SafeHandle, Byte*, Int32)
00007F12E5082560 00007F12E660FADB ILStubClass.IL_STUB_PInvoke(System.Runtime.InteropServices.SafeHandle, Byte*, Int32)
00007F12E5082600 00007F12E74AFE15 System.IO.FileSystemWatcher+RunningInstance.TryReadEvent(NotifyEvent ByRef)
00007F12E5082650 00007F12E74AFBEE System.IO.FileSystemWatcher+RunningInstance.ProcessEvents()
00007F12E5082730 00007F12E74B02E5 System.IO.FileSystemWatcher+RunningInstance+<>c.<Start>b__14_0(System.Object)
00007F12E5082740 00007F12E617E9A3 System.Threading.Tasks.Task.InnerInvoke()
00007F12E5082760 00007F12E6184891 System.Threading.Tasks.Task+<>c.<.cctor>b__277_0(System.Object)
00007F12E5082770 00007F12E6169841 System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
00007F12E50827C0 00007F12E617E79A System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
00007F12E5082840 00007F12E617E598 System.Threading.Tasks.Task.ExecuteEntryUnsafe(System.Threading.Thread)
00007F12E5082860 00007F12E618BCB7 System.Threading.Tasks.ThreadPoolTaskScheduler+<>c.<.cctor>b__10_0(System.Object)
00007F12E5082870 00007F12E616045F System.Threading.ThreadHelper.ThreadStart_Context(System.Object)
00007F12E5082890 00007F12E6169841 System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
00007F12E50828E0 00007F12E61604F1 System.Threading.ThreadHelper.ThreadStart(System.Object)
00007F12E5082CE0 00007f135fab9b47 [DebuggerU2MCatchHandlerFrame: 00007f12e5082ce0]
OS Thread Id: 0x10
        Child SP               IP Call Site
00007F12E4880940 00007f136080400c [HelperMethodFrame_1OBJ: 00007f12e4880940] System.Threading.Monitor.ObjWait(Boolean, Int32, System.Object)
00007F12E4880A70 00007F12EC85400A System.Threading.ManualResetEventSlim.Wait(Int32, System.Threading.CancellationToken)
00007F12E4880AF0 00007F12EC853AE2 System.Threading.Tasks.Task.SpinThenBlockingWait(Int32, System.Threading.CancellationToken)
00007F12E4880B50 00007F12EC8538B0 System.Threading.Tasks.Task.InternalWaitCore(Int32, System.Threading.CancellationToken)
00007F12E4880BA0 00007F12EB151E14 /app/APPNAME.BusinessLogic.dll!Unknown
00007F12E4880CF0 00007F12EB150C7D /app/APPNAME.Jobs.dll!Unknown
00007F12E4880E00 00007F12EB1505A7 /app/APPNAME.Jobs.dll!Unknown
00007F12E4880EF0 00007F12EB1503A4 /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Private.CoreLib.dll!Unknown
00007F12E4880F60 00007F12EB1502FB /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Private.CoreLib.dll!Unknown
00007F12E4880F80 00007F12EB1502B6 /app/APPNAME.Jobs.dll!Unknown
00007F12E4880FE0 00007F12EB1501F7 /app/APPNAME.Quartz.dll!Unknown
00007F12E4881010 00007F12EB14F505 /app/APPNAME.Quartz.dll!Unknown
00007F12E48811D0 00007F12EB14EFB4 /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Private.CoreLib.dll!Unknown
00007F12E4881240 00007F12EB14EF0B /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Private.CoreLib.dll!Unknown
00007F12E4881260 00007F12EB14EEC2 /app/APPNAME.Quartz.dll!Unknown
00007F12E48812D0 00007F12EB14A0B4 /app/Quartz.dll!Unknown
00007F12E48815E0 00007F12EB149324 /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Private.CoreLib.dll!Unknown
00007F12E4881650 00007F12EB14927B /usr/share/dotnet/shared/Microsoft.NETCore.App/5.0.4/System.Private.CoreLib.dll!Unknown
00007F12E4881670 00007F12EB149231 /app/Quartz.dll!Unknown
00007F12E4881730 00007F12EB149173 /app/Quartz.dll!Unknown
00007F12E4881760 00007F12E61773E1 System.Threading.Tasks.Task`1[[System.__Canon, System.Private.CoreLib]].InnerInvoke()
00007F12E4881780 00007F12E6184891 System.Threading.Tasks.Task+<>c.<.cctor>b__277_0(System.Object)
00007F12E4881790 00007F12E6169992 System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(System.Threading.Thread, System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
00007F12E48817D0 00007F12E617E7B9 System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
00007F12E4881850 00007F12E617E598 System.Threading.Tasks.Task.ExecuteEntryUnsafe(System.Threading.Thread)
00007F12E4881870 00007F12E617E537 System.Threading.Tasks.Task.ExecuteFromThreadPool(System.Threading.Thread)
00007F12E4881880 00007F12E6172B73 System.Threading.ThreadPoolWorkQueue.Dispatch()
00007F12E4881900 00007F12E6161E67 System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
00007F12E4881C80 00007f135fab9b47 [DebuggerU2MCatchHandlerFrame: 00007f12e4881c80]
OS Thread Id: 0x11
        Child SP               IP Call Site
00007F118FA61570 00007f1360807544 [InlinedCallFrame: 00007f118fa61570] Interop+Sys.Read(System.Runtime.InteropServices.SafeHandle, Byte*, Int32)
00007F118FA61570 00007f12e660fadb [InlinedCallFrame: 00007f118fa61570] Interop+Sys.Read(System.Runtime.InteropServices.SafeHandle, Byte*, Int32)
00007F118FA61560 00007F12E660FADB ILStubClass.IL_STUB_PInvoke(System.Runtime.InteropServices.SafeHandle, Byte*, Int32)
00007F118FA61600 00007F12E74AFE15 System.IO.FileSystemWatcher+RunningInstance.TryReadEvent(NotifyEvent ByRef)
00007F118FA61650 00007F12E74AFBEE System.IO.FileSystemWatcher+RunningInstance.ProcessEvents()
00007F118FA61730 00007F12E74B02E5 System.IO.FileSystemWatcher+RunningInstance+<>c.<Start>b__14_0(System.Object)
00007F118FA61740 00007F12E617E9A3 System.Threading.Tasks.Task.InnerInvoke()
00007F118FA61760 00007F12E6184891 System.Threading.Tasks.Task+<>c.<.cctor>b__277_0(System.Object)
00007F118FA61770 00007F12E6169841 System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
00007F118FA617C0 00007F12E617E79A System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
00007F118FA61840 00007F12E617E598 System.Threading.Tasks.Task.ExecuteEntryUnsafe(System.Threading.Thread)
00007F118FA61860 00007F12E618BCB7 System.Threading.Tasks.ThreadPoolTaskScheduler+<>c.<.cctor>b__10_0(System.Object)
00007F118FA61870 00007F12E616045F System.Threading.ThreadHelper.ThreadStart_Context(System.Object)
00007F118FA61890 00007F12E6169841 System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
00007F118FA618E0 00007F12E61604F1 System.Threading.ThreadHelper.ThreadStart(System.Object)
00007F118FA61CE0 00007f135fab9b47 [DebuggerU2MCatchHandlerFrame: 00007f118fa61ce0]
OS Thread Id: 0x12
        Child SP               IP Call Site
00007F118F1C3800 00007f13604087ef [InlinedCallFrame: 00007f118f1c3800] Interop+Sys.WaitForSocketEvents(IntPtr, SocketEvent*, Int32*)
00007F118F1C3800 00007f12ea08a8d8 [InlinedCallFrame: 00007f118f1c3800] Interop+Sys.WaitForSocketEvents(IntPtr, SocketEvent*, Int32*)
00007F118F1C37F0 00007F12EA08A8D8 ILStubClass.IL_STUB_PInvoke(IntPtr, SocketEvent*, Int32*)
00007F118F1C3880 00007F12E8CF9174 System.Net.Sockets.SocketAsyncEngine.EventLoop()
00007F118F1C38D0 00007F12E8CF96E5 System.Net.Sockets.SocketAsyncEngine+<>c.<.ctor>b__14_0(System.Object)
00007F118F1C38E0 00007F12E616051A System.Threading.ThreadHelper.ThreadStart(System.Object)
00007F118F1C3CE0 00007f135fab9b47 [DebuggerU2MCatchHandlerFrame: 00007f118f1c3ce0]
OS Thread Id: 0x13
        Child SP               IP Call Site
00007F118E4706E8 00007f13608043f9 [HelperMethodFrame: 00007f118e4706e8] System.Threading.Thread.SleepInternal(Int32)
00007F118E470830 00007F12EC8681F0 System.Threading.Thread.Sleep(System.TimeSpan)
00007F118E470840 00007F12E6F10C29 Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.Heartbeat.TimerLoop()
00007F118E470860 00007F12E6F60DB5 Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.Heartbeat+<>c.<.ctor>b__8_0(System.Object)
00007F118E470870 00007F12E616045F System.Threading.ThreadHelper.ThreadStart_Context(System.Object)
00007F118E470890 00007F12E6169841 System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
00007F118E4708E0 00007F12E61604F1 System.Threading.ThreadHelper.ThreadStart(System.Object)
00007F118E470CE0 00007f135fab9b47 [DebuggerU2MCatchHandlerFrame: 00007f118e470ce0]
OS Thread Id: 0x21e
        Child SP               IP Call Site
OS Thread Id: 0x229
        Child SP               IP Call Site
OS Thread Id: 0x22a
        Child SP               IP Call Site
OS Thread Id: 0x22b
        Child SP               IP Call Site
```

**Comment by Ne4to** [comment]
> What is the type of _httpClientHandler here?

It is inheritor of HttpClientHandler with helper methods.

> Getting a network trace would help to confirm or refute that.

I was able to run tcpdump, now waiting for a freeze. Looks like regular responses have valid endings: `0d 0a 30 0d 0a 0d 0a` (`\r\n0\r\n\r\n`)

**Comment by karelz** [comment]
@Ne4to were you able to find out anything from the dump? It seems this is not very actionable for us.
If you bring back HttpClient, or custom timeouts, we expect the key problems you are facing should go away ...

**Comment by Ne4to** [comment]
HttpClientHandler was replaced by HttpClient in generated code. It is working for 2 weeks now and no issues so far. I'd like to give it a little more time and wait to the end of the week. 

One of the differences between custom implementation and HttpClient I found is `LoadIntoBufferAsync` call. HttpClient calls internal overload of [LoadIntoBufferAsync](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs#L534) with CancellationToken. Custom implementation used public method without this parameter so `CancellationToken.None` was used.

Custom implementation:
```
await response.Content.LoadIntoBufferAsync().ConfigureAwait(false);
```
HttpClient:
```
await response.Content.LoadIntoBufferAsync(_maxResponseContentBufferSize, cts.Token).ConfigureAwait(false);
```
